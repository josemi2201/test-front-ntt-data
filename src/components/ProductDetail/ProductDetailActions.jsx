import { mdiCartArrowDown } from '@mdi/js';
import { Box, Typography } from '@mui/material';
import { useFormik } from 'formik';
import PropTypes from 'prop-types';
import React from 'react';
import * as yup from 'yup';
import { showToast } from '../../helpers/toast';
import { useData } from '../../hooks/useData';
import { Button } from '../UI/Button';
import { Select } from '../UI/Select';


const validationSchema = yup.object({
  storageCode: yup
      .string()
      .required('Color is required'),
  colorCode: yup
      .string()
      .required('Storage is required'),
});

export const ProductDetailActions = ({product}) => {

  const { addCart } = useData();
  
  const { values, errors, touched, isSubmitting, handleSubmit, handleChange } = useFormik({
    initialValues: {
      storageCode: '',
      colorCode: '',
    },
    validationSchema: validationSchema,
    onSubmit: async (values, { setSubmitting }) => {
        try {    
            const data = {
              id: product.id,
              ...values,
            }

            await addCart(data)

            setSubmitting(false)
            
            showToast("Product added to cart", "success")
        } catch (error) {
          showToast(error.message, "error")
        }
    },
  });

  return (
    <Box>
      <Typography
        variant="h4"
        sx={sx.title}
      >
        Options
      </Typography>
      <Box component="hr" />
      <Box>
        <Box
          component="form"
          onSubmit={handleSubmit}
          sx={sx.form}
        >
          <Select 
            label="Storage"
            name="storageCode"
            options={product.storages}
            value={values.storageCode}
            handleChange={handleChange}
            touched={touched}
            errors={errors}
          />
          <Select 
            label="Color"
            name="colorCode"
            options={product.colors}
            value={values.colorCode}
            handleChange={handleChange}
            touched={touched}
            errors={errors}
          />
          <Button 
            icon={mdiCartArrowDown}
            text="Add to cart"
            type="submit"
            isLoading={isSubmitting}
          />
        </Box>
      </Box>
    </Box>
  )
}

const sx = {
  form:{
    display: 'flex',
    flexDirection: 'column',
    gap: 3,
    marginTop: 3,
  }
}

ProductDetailActions.propTypes = {
  product: PropTypes.object.isRequired
}