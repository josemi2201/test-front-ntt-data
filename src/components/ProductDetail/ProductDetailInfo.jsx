import { Box, Typography, useTheme } from "@mui/material";
import PropTypes from "prop-types";
import React from "react";
import { PropertyView } from "../UI/PropertyView";

export const ProductDetailInfo = ({product}) => {

  const {
    brand,
    model,
    priceDescription,
    processor,
    ram,
    so,
    screenResolution,
    battery,
    cameras,
    dimensions,
    weight
  } = product;

  const { palette: { secondary } } = useTheme();

  return (
    <Box>
      <Typography
        variant="h4"
      >
        {model}
      </Typography>
      <Typography
        sx={{
          color: secondary.light
        }}
      >
        {brand}
      </Typography>
      <Typography
        variant="h5"
        sx={sx.price}
      >
        {priceDescription}
      </Typography>

      <Box component="hr" />
      <Box
        component="ul"
      >
        <PropertyView
          description="model"
          value={model}
        />
        <PropertyView
          description="processor"
          value={processor}
        />
        <PropertyView
          description="RAM"
          value={ram}
        />
        <PropertyView
          description="S.O."
          value={so}
        />
        <PropertyView
          description="screen resolution"
          value={screenResolution}
        />
        <PropertyView
          description="battery"
          value={battery}
        />
        <PropertyView
          description="cameras"
          value={cameras}
        />
        <PropertyView
          description="dimensions"
          value={dimensions}
        />
        <PropertyView
          description="weight"
          value={weight}
        />
      </Box>
    </Box>
  );

};

const sx = {
  price: {
    fontWeight: "bold"
  }
};

ProductDetailInfo.propTypes = {
  product: PropTypes.object.isRequired
};
