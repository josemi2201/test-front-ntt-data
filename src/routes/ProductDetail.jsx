import { Box, Grid } from '@mui/material'
import React, { useEffect, useState } from 'react'
import { useParams } from 'react-router-dom'
import { Header } from '../components/Header/Header'
import { ProductDetailActions } from '../components/ProductDetail/ProductDetailActions'
import { ProductDetailInfo } from '../components/ProductDetail/ProductDetailInfo'
import { useData } from '../hooks/useData'

export const ProductDetail = () => {

  const { getProduct } = useData();

  const { id } = useParams()
  const [product, setProduct] = useState({})

  const { 
    imgUrl,
  } = product

  useEffect(() => {
    getProduct(id)
      .then( product => setProduct(product))
      .catch( error => alert(error.message))
  }, [])

  return (
    <Header>
      <Grid
        container
        sx={sx.contProduct}
      >
        <Grid
          item
          xs={6}
          sx={sx.contImg}
          component="section"
          >
          <Box
            component="img"
            src={imgUrl}
            sx={sx.img}
          />
        </Grid>
        <Grid
          item
          xs={6}
          sx={sx.gridInfo}
          component="section"
        >
          <Box
            sx={sx.module}
          >
            <ProductDetailInfo product={product} />
          </Box>
          <Box
            sx={sx.module}
          >
            <ProductDetailActions product={product} />
          </Box>
        </Grid>
      </Grid>
    </Header>
  )
}

const PADDING_INFO = 32

const sx = {
  contProduct: {
    height: "100%",
  },
  img: {
    width: "100%",
    borderRadius: 2,
  },
  gridInfo: {
    display: "flex",
    flexDirection: "column",
    gap: 4,
  },
  module:{
    border: "1px solid grey",
    borderRadius: 2,
    margin: "0 32px",
    padding: `${PADDING_INFO}px`,
    backgroundColor: "white",
  }
}

