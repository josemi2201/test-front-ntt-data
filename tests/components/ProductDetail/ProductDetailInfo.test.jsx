import { render } from '@testing-library/react';
import { ProductDetailInfo } from '../../../src/components/ProductDetail/ProductDetailInfo';

describe("Test <ProductDetailInfo />", () => {
  test('Match whit snapshot', () => {
    const product =  {
        id: "AasKFs5EGbyAEIKkcHQcF",
        brand: "alcatel",
        model: "Flash (2017)",
        price: "",
        imgUrl: "https://www.gizbot.com/imgh/310x233/2020/09/realme-narzo-20_160145525500.jpg",
        processor: "MediaTek MT8735",
        ram: "16GB | 32GB | 64GB",
        so: "Android 10",
        screenResolution: "1280x720",
        battery: "3.400mAh",
        cameras: "13MP | 5MP",
        dimensions: "99,8 mm / 190 mm / 8,5 mm",
        weight: "275 g",
        storages: [
            { id: "1", label: "16GB" },
            { id: "2", label: "32GB" },
            { id: "3", label: "64GB" },
            { id: "4", label: "128GB" },
        ],
        colors: [{ id: "1", label: "Black" }],
    }

    const { container } = render( <ProductDetailInfo product={product} /> );
    expect( container ).toMatchSnapshot();

  });
})