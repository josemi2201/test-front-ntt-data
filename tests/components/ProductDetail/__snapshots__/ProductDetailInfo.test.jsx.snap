// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <ProductDetailInfo /> Match whit snapshot 1`] = `
<div>
  <div
    class="MuiBox-root css-0"
  >
    <h4
      class="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
    >
      Flash (2017)
    </h4>
    <p
      class="MuiTypography-root MuiTypography-body1 css-sc70jj-MuiTypography-root"
    >
      alcatel
    </p>
    <h5
      class="MuiTypography-root MuiTypography-h5 css-bezmlo-MuiTypography-root"
    />
    <hr
      class="MuiBox-root css-0"
    />
    <ul
      class="MuiBox-root css-0"
    >
      <li
        class="MuiBox-root css-0"
      >
        <div
          class="MuiBox-root css-k008qs"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-c7zlro-MuiTypography-root"
          >
            model: 
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-yyvcc2-MuiTypography-root"
          >
            Flash (2017)
          </p>
        </div>
      </li>
      <li
        class="MuiBox-root css-0"
      >
        <div
          class="MuiBox-root css-k008qs"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-c7zlro-MuiTypography-root"
          >
            processor: 
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-yyvcc2-MuiTypography-root"
          >
            MediaTek MT8735
          </p>
        </div>
      </li>
      <li
        class="MuiBox-root css-0"
      >
        <div
          class="MuiBox-root css-k008qs"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-c7zlro-MuiTypography-root"
          >
            RAM: 
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-yyvcc2-MuiTypography-root"
          >
            16GB | 32GB | 64GB
          </p>
        </div>
      </li>
      <li
        class="MuiBox-root css-0"
      >
        <div
          class="MuiBox-root css-k008qs"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-c7zlro-MuiTypography-root"
          >
            S.O.: 
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-yyvcc2-MuiTypography-root"
          >
            Android 10
          </p>
        </div>
      </li>
      <li
        class="MuiBox-root css-0"
      >
        <div
          class="MuiBox-root css-k008qs"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-c7zlro-MuiTypography-root"
          >
            screen resolution: 
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-yyvcc2-MuiTypography-root"
          >
            1280x720
          </p>
        </div>
      </li>
      <li
        class="MuiBox-root css-0"
      >
        <div
          class="MuiBox-root css-k008qs"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-c7zlro-MuiTypography-root"
          >
            battery: 
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-yyvcc2-MuiTypography-root"
          >
            3.400mAh
          </p>
        </div>
      </li>
      <li
        class="MuiBox-root css-0"
      >
        <div
          class="MuiBox-root css-k008qs"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-c7zlro-MuiTypography-root"
          >
            cameras: 
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-yyvcc2-MuiTypography-root"
          >
            13MP | 5MP
          </p>
        </div>
      </li>
      <li
        class="MuiBox-root css-0"
      >
        <div
          class="MuiBox-root css-k008qs"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-c7zlro-MuiTypography-root"
          >
            dimensions: 
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-yyvcc2-MuiTypography-root"
          >
            99,8 mm / 190 mm / 8,5 mm
          </p>
        </div>
      </li>
      <li
        class="MuiBox-root css-0"
      >
        <div
          class="MuiBox-root css-k008qs"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-c7zlro-MuiTypography-root"
          >
            weight: 
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-yyvcc2-MuiTypography-root"
          >
            275 g
          </p>
        </div>
      </li>
    </ul>
  </div>
</div>
`;
